In [2]:
import pandas as pd
import numpy as np

import scipy.sparse
import sklearn.feature_extraction

import matplotlib.pylab as plt
%matplotlib inline

from tqdm import tqdm
import platform

pd.set_option("display.max_rows", 10)
pd.set_option('display.max_columns', 1100)

import os

%pylab inline
warnings.filterwarnings('ignore')

Populating the interactive namespace from numpy and matplotlib


### Load vectorizers

In [22]:
infiles = ['parted-cls/orders_vectorizer30.joblib.pkl.0']

import os
from sklearn.externals import joblib
from lib.helpers.fileSplitter import cat_files

cat_files(infiles, 'cls/orders_vectorizer30.joblib.pkl')

filename = 'cls/orders_vectorizer30.joblib.pkl'
orders_vectorizer = joblib.load(filename)

infiles = ['parted-cls/values_vectorizer30.joblib.pkl.0']

cat_files(infiles, 'cls/values_vectorizer30.joblib.pkl')

filename = 'cls/values_vectorizer30.joblib.pkl'
values_vectorizer = joblib.load(filename)

### Load classifyers 

In [4]:
infiles = ['parted-cls/device_class_code_logreg_cls.joblib.pkl.0']

cat_files(infiles, 'cls/device_class_code_logreg_cls.joblib.pkl')
clf_device_class_code = joblib.load('cls/device_class_code_logreg_cls.joblib.pkl')

infiles = ['parted-cls/os_family_code_logreg_cls.joblib.pkl.0']

cat_files(infiles, 'cls/os_family_code_logreg_cls.joblib.pkl')
clf_os_family_code = joblib.load('cls/os_family_code_logreg_cls.joblib.pkl')

infiles = ['parted-cls/os_code_logreg_cls.joblib.pkl.0']

cat_files(infiles, 'cls/os_code_logreg_cls.joblib.pkl')
clf_os_code = joblib.load('cls/os_code_logreg_cls.joblib.pkl')

infiles = ['parted-cls/ua_class_code_logreg_cls.joblib.pkl.0']

cat_files(infiles, 'cls/ua_class_code_logreg_cls.joblib.pkl')
clf_ua_class_code = joblib.load('cls/ua_class_code_logreg_cls.joblib.pkl')

infiles = ['parted-cls/ua_family_code_logreg_cls.joblib.pkl.0']

cat_files(infiles, 'cls/ua_family_code_logreg_cls.joblib.pkl')
clf_ua_family_code = joblib.load('cls/ua_family_code_logreg_cls.joblib.pkl')

infiles = ['parted-cls/ua_version_logreg_cls.joblib.pkl.0']

cat_files(infiles, 'cls/ua_version_logreg_cls.joblib.pkl')
clf_ua_version = joblib.load('cls/ua_version_logreg_cls.joblib.pkl')

### Load test data

In [18]:
main_data = np.load('df/main2_data.npy').tolist()[:20000]
values_data = np.load('df/values2_data.npy').tolist()[:20000]
order_data = np.load('df/order2_data.npy').tolist()[:20000]

main_df = pd.DataFrame(main_data)

main_df

User_Agent device_class_code  \
0      Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKi...           desktop   
1      Mozilla/5.0 (Windows NT 6.3; WOW64; rv:34.0) G...           desktop   
2      Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.3...           desktop   
3      Mozilla/5.0 (Windows NT 5.1; rv:7.0.1) Gecko/2...           desktop   
4      Mozilla/5.0 (Linux; Android 6.0.1; SAMSUNG SM-...        smartphone   
...                                                  ...               ...   
19995  Mozilla/5.0 (Windows NT 10.0) AppleWebKit/537....           desktop   
19996  Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebK...           desktop   
19997  Mozilla/5.0 (Windows NT 6.1; rv:39.0) Gecko/20...           desktop   
19998  Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKi...           desktop   
19999  Mozilla/5.0 (Windows NT 6.1; rv:39.0) Gecko/20...           desktop   

                   ip      os_code os_family_code   timestamp   ua_class_code  \
0         62.89.6.193    windows_7        windows  1488578454         browser   
1        62.210.80.74  windows_8_1        windows  1488578468         browser   
2       77.87.101.132    windows_7        windows  1488578472         browser   
3       216.239.90.19   windows_xp        windows  1488578508         browser   
4         37.19.72.34    android_6        android  1488578544  mobile_browser   
...               ...          ...            ...         ...             ...   
19995    85.26.232.62   windows_10        windows  1488703450         browser   
19996    46.118.53.27   windows_10        windows  1488703452         browser   
19997   195.70.209.78    windows_7        windows  1488703453         browser   
19998  37.151.196.120    windows_7        windows  1488703458         browser   
19999   195.70.209.78    windows_7        windows  1488703458         browser   

               ua_family_code                 ua_version  
0                      chrome         chrome56.0.2924.87  
1                     firefox                firefox34.0  
2              yandex-browser  yandex-browser17.1.1.1003  
3                     firefox               firefox7.0.1  
4      mobile_samsung_browser  mobile_samsung_browser4.0  
...                       ...                        ...  
19995                  chrome         chrome56.0.2924.87  
19996          yandex-browser  yandex-browser17.3.0.1785  
19997                 firefox                firefox39.0  
19998                   amigo         amigo54.0.2840.191  
19999                 firefox                firefox39.0  

[20000 rows x 9 columns]

In [23]:
important_values_keys_set = {
    'Accept', 
    'Accept-Charset', 
    'Accept-Encoding'
}
important_orders_keys_set = {
    'Upgrade-Insecure-Requests',
    'Accept', 
    'If-Modified-Since',
    'Host', 
    'Connection', 
    'User-Agent', 
    'From', 
    'Accept-Encoding' 
}

from lib.parsers.logParser import LogParser

l_parser = LogParser(log_folder='Logs/')

l_parser.reassign_orders_values(order_data, values_data)
X_test = l_parser.prepare_data(orders_vectorizer, values_vectorizer, important_orders_keys_set, important_values_keys_set, fit_dict=False)

100%|██████████| 20000/20000 [00:00<00:00, 230988.68it/s]

Sparse dummy orders shape: 
(20000, 50)
Sparse dummy values shape: 
(20000, 516)


### Calculate scores

Примечание: Для Decision Tree в cross_val_score по умолчанию берется показатель 'Accuracy'

Поскольку 'Accuracy' для линейной регрессии линейный мы не будем считать на 3-х или 5-ти фолдах(долго), а просто возьмем от тренировочной выборки 'Accuracy'

In [132]:
thres = 0.01

**Browser (clf_ua_family_code)**

In [133]:
from lib.thresholdPredictions import ThresholdPredictions

pred = ThresholdPredictions(user_agent_list=clf_ua_family_code.classes_.tolist(), clf=clf_ua_family_code)
y_test_names, y_predicted, compare_answers, answers_count = pred.predict(X_test, main_df.ua_family_code, thres, sparce_y=False, single_labels=True)

20000it [00:01, 16369.04it/s]


In [134]:
compare_frame = pd.concat(
    [
        pd.DataFrame(y_test_names),
        y_predicted, 
        pd.DataFrame(compare_answers),  
        pd.DataFrame(answers_count)
    ], keys=['browser_name', 'predicted_browser_name', 'browser_name_correctness', 'browser_name_count'], axis=1, join='inner')

compare_frame

browser_name  \
               ua_family_code   
0                      chrome   
1                     firefox   
2              yandex-browser   
3                     firefox   
4      mobile_samsung_browser   
...                       ...   
19995                  chrome   
19996          yandex-browser   
19997                 firefox   
19998                   amigo   
19999                 firefox   

                                  predicted_browser_name  \
                                                       0   
0      [amigo, chrome, chrome_mobile, opera, yandex-b...   
1      [chrome, chrome_mobile, chromodo, firefox, ie,...   
2      [amigo, chrome, chrome_mobile, opera, yandex-b...   
3                                              [firefox]   
4      [amigo, chrome, chrome_mobile, mobile_samsung_...   
...                                                  ...   
19995  [amigo, chrome, chrome_mobile, opera, yandex-b...   
19996  [amigo, chrome, chrome_mobile, opera, yandex-b...   
19997  [chrome, firefox, firefox_mobile, safari, safa...   
19998  [amigo, chrome, chrome_mobile, opera, yandex-b...   
19999  [chrome, firefox, firefox_mobile, safari, safa...   

      browser_name_correctness browser_name_count  
                             0                  0  
0                         True                  6  
1                         True                 11  
2                         True                  6  
3                         True                  1  
4                         True                  6  
...                        ...                ...  
19995                     True                  6  
19996                     True                  6  
19997                     True                  5  
19998                     True                  6  
19999                     True                  5  

[20000 rows x 4 columns]

Accuracy: $ACC = \frac{TP + TN}{P + N},\ \ \mathrm{where}\ \ P + N = length,\ \ TP = sum(True), \ \ TN = 0$

In [135]:
print('Accuracy: {}'.format(sum(compare_frame.browser_name_correctness[0])/20000))

Accuracy: 0.96125


**Browser + Browser version (clf_ua_family_code + clf_ua_version)**

In [136]:
pred = ThresholdPredictions(user_agent_list=clf_ua_version.classes_.tolist(), clf=clf_ua_version)
y_test_names, y_predicted, compare_answers, answers_count = pred.predict(X_test, main_df.ua_version, thres, sparce_y=False, single_labels=True)

20000it [00:00, 35208.96it/s]


In [137]:
compare_frame['browser_version'] = pd.DataFrame(y_test_names)
compare_frame['predicted_browser_version'] = y_predicted
compare_frame['browser_version_correctness'] = pd.DataFrame(compare_answers)
compare_frame['browser_version_count'] = pd.DataFrame(answers_count)

In [138]:
compare_frame

browser_name  \
               ua_family_code   
0                      chrome   
1                     firefox   
2              yandex-browser   
3                     firefox   
4      mobile_samsung_browser   
...                       ...   
19995                  chrome   
19996          yandex-browser   
19997                 firefox   
19998                   amigo   
19999                 firefox   

                                  predicted_browser_name  \
                                                       0   
0      [amigo, chrome, chrome_mobile, opera, yandex-b...   
1      [chrome, chrome_mobile, chromodo, firefox, ie,...   
2      [amigo, chrome, chrome_mobile, opera, yandex-b...   
3                                              [firefox]   
4      [amigo, chrome, chrome_mobile, mobile_samsung_...   
...                                                  ...   
19995  [amigo, chrome, chrome_mobile, opera, yandex-b...   
19996  [amigo, chrome, chrome_mobile, opera, yandex-b...   
19997  [chrome, firefox, firefox_mobile, safari, safa...   
19998  [amigo, chrome, chrome_mobile, opera, yandex-b...   
19999  [chrome, firefox, firefox_mobile, safari, safa...   

      browser_name_correctness browser_name_count            browser_version  \
                             0                  0                              
0                         True                  6         chrome56.0.2924.87   
1                         True                 11                firefox34.0   
2                         True                  6  yandex-browser17.1.1.1003   
3                         True                  1               firefox7.0.1   
4                         True                  6  mobile_samsung_browser4.0   
...                        ...                ...                        ...   
19995                     True                  6         chrome56.0.2924.87   
19996                     True                  6  yandex-browser17.3.0.1785   
19997                     True                  5                firefox39.0   
19998                     True                  6         amigo54.0.2840.191   
19999                     True                  5                firefox39.0   

                               predicted_browser_version  \
                                                           
0      [chrome55.0.2883.87, chrome56.0.2924.87, chrom...   
1      [chrome36.0.1985.125, chrome39.0.2171.71, chro...   
2      [chrome55.0.2883.87, chrome56.0.2924.87, chrom...   
3                                         [firefox7.0.1]   
4      [amigo47.5.2526.115, chrome49.0.2623.112, chro...   
...                                                  ...   
19995  [chrome55.0.2883.87, chrome56.0.2924.87, chrom...   
19996  [chrome55.0.2883.87, chrome56.0.2924.87, chrom...   
19997  [firefox12.0, firefox15.0.1, firefox38.0, fire...   
19998  [chrome55.0.2883.87, chrome56.0.2924.87, chrom...   
19999  [firefox12.0, firefox15.0.1, firefox38.0, fire...   

      browser_version_correctness browser_version_count  
                                                         
0                            True                    12  
1                            True                    19  
2                            True                    12  
3                            True                     1  
4                            True                    11  
...                           ...                   ...  
19995                        True                    12  
19996                       False                    12  
19997                       False                    13  
19998                       False                    12  
19999                       False                    13  

[20000 rows x 8 columns]

In [139]:
print('Browser version accuracy: {}'.format(sum(compare_frame.browser_version_correctness)/20000))
print('Conditional accuracy: {}'.format(sum(compare_frame.browser_name_correctness[0] & compare_frame.browser_version_correctness)/20000))

Browser version accuracy: 0.8155
Conditional accuracy: 0.8141


**Browser + Browser version + Platform (clf_ua_family_code + clf_ua_version + clf_os_family_code)**

In [140]:
pred = ThresholdPredictions(user_agent_list=clf_os_family_code.classes_.tolist(), clf=clf_os_family_code)
y_test_names, y_predicted, compare_answers, answers_count = pred.predict(X_test, main_df.os_family_code, thres, sparce_y=False, single_labels=True)

20000it [00:00, 36212.81it/s]


In [141]:
compare_frame['platform'] = pd.DataFrame(y_test_names)
compare_frame['predicted_platform'] = y_predicted
compare_frame['platform_correctness'] = pd.DataFrame(compare_answers)
compare_frame['platform_count'] = pd.DataFrame(answers_count)
compare_frame

browser_name  \
               ua_family_code   
0                      chrome   
1                     firefox   
2              yandex-browser   
3                     firefox   
4      mobile_samsung_browser   
...                       ...   
19995                  chrome   
19996          yandex-browser   
19997                 firefox   
19998                   amigo   
19999                 firefox   

                                  predicted_browser_name  \
                                                       0   
0      [amigo, chrome, chrome_mobile, opera, yandex-b...   
1      [chrome, chrome_mobile, chromodo, firefox, ie,...   
2      [amigo, chrome, chrome_mobile, opera, yandex-b...   
3                                              [firefox]   
4      [amigo, chrome, chrome_mobile, mobile_samsung_...   
...                                                  ...   
19995  [amigo, chrome, chrome_mobile, opera, yandex-b...   
19996  [amigo, chrome, chrome_mobile, opera, yandex-b...   
19997  [chrome, firefox, firefox_mobile, safari, safa...   
19998  [amigo, chrome, chrome_mobile, opera, yandex-b...   
19999  [chrome, firefox, firefox_mobile, safari, safa...   

      browser_name_correctness browser_name_count            browser_version  \
                             0                  0                              
0                         True                  6         chrome56.0.2924.87   
1                         True                 11                firefox34.0   
2                         True                  6  yandex-browser17.1.1.1003   
3                         True                  1               firefox7.0.1   
4                         True                  6  mobile_samsung_browser4.0   
...                        ...                ...                        ...   
19995                     True                  6         chrome56.0.2924.87   
19996                     True                  6  yandex-browser17.3.0.1785   
19997                     True                  5                firefox39.0   
19998                     True                  6         amigo54.0.2840.191   
19999                     True                  5                firefox39.0   

                               predicted_browser_version  \
                                                           
0      [chrome55.0.2883.87, chrome56.0.2924.87, chrom...   
1      [chrome36.0.1985.125, chrome39.0.2171.71, chro...   
2      [chrome55.0.2883.87, chrome56.0.2924.87, chrom...   
3                                         [firefox7.0.1]   
4      [amigo47.5.2526.115, chrome49.0.2623.112, chro...   
...                                                  ...   
19995  [chrome55.0.2883.87, chrome56.0.2924.87, chrom...   
19996  [chrome55.0.2883.87, chrome56.0.2924.87, chrom...   
19997  [firefox12.0, firefox15.0.1, firefox38.0, fire...   
19998  [chrome55.0.2883.87, chrome56.0.2924.87, chrom...   
19999  [firefox12.0, firefox15.0.1, firefox38.0, fire...   

      browser_version_correctness browser_version_count platform  \
                                                                   
0                            True                    12  windows   
1                            True                    19  windows   
2                            True                    12  windows   
3                            True                     1  windows   
4                            True                    11  android   
...                           ...                   ...      ...   
19995                        True                    12  windows   
19996                       False                    12  windows   
19997                       False                    13  windows   
19998                       False                    12  windows   
19999                       False                    13  windows   

                                   predicted_platform platform_correctness  \
                     

In [142]:
print('Platform accuracy: {}'.format(sum(compare_frame.platform_correctness)/20000))
print('Conditional accuracy: {}'.format(sum(compare_frame.browser_name_correctness[0] & compare_frame.browser_version_correctness & compare_frame.platform_correctness)/20000))

Platform accuracy: 0.9648
Conditional accuracy: 0.80905


**Browser + Browser version + Platform + Platform version (clf_ua_family_code + clf_ua_version + clf_os_family_code + clf_os_code)**

In [143]:
pred = ThresholdPredictions(user_agent_list=clf_os_code.classes_.tolist(), clf=clf_os_code)
y_test_names, y_predicted, compare_answers, answers_count = pred.predict(X_test, main_df.os_code, thres, sparce_y=False, single_labels=True)

20000it [00:00, 34139.20it/s]


In [144]:
compare_frame['platform_version'] = pd.DataFrame(y_test_names)
compare_frame['predicted_platform_version'] = y_predicted
compare_frame['platform_version_correctness'] = pd.DataFrame(compare_answers)
compare_frame['platform_version_count'] = pd.DataFrame(answers_count)
compare_frame

browser_name  \
               ua_family_code   
0                      chrome   
1                     firefox   
2              yandex-browser   
3                     firefox   
4      mobile_samsung_browser   
...                       ...   
19995                  chrome   
19996          yandex-browser   
19997                 firefox   
19998                   amigo   
19999                 firefox   

                                  predicted_browser_name  \
                                                       0   
0      [amigo, chrome, chrome_mobile, opera, yandex-b...   
1      [chrome, chrome_mobile, chromodo, firefox, ie,...   
2      [amigo, chrome, chrome_mobile, opera, yandex-b...   
3                                              [firefox]   
4      [amigo, chrome, chrome_mobile, mobile_samsung_...   
...                                                  ...   
19995  [amigo, chrome, chrome_mobile, opera, yandex-b...   
19996  [amigo, chrome, chrome_mobile, opera, yandex-b...   
19997  [chrome, firefox, firefox_mobile, safari, safa...   
19998  [amigo, chrome, chrome_mobile, opera, yandex-b...   
19999  [chrome, firefox, firefox_mobile, safari, safa...   

      browser_name_correctness browser_name_count            browser_version  \
                             0                  0                              
0                         True                  6         chrome56.0.2924.87   
1                         True                 11                firefox34.0   
2                         True                  6  yandex-browser17.1.1.1003   
3                         True                  1               firefox7.0.1   
4                         True                  6  mobile_samsung_browser4.0   
...                        ...                ...                        ...   
19995                     True                  6         chrome56.0.2924.87   
19996                     True                  6  yandex-browser17.3.0.1785   
19997                     True                  5                firefox39.0   
19998                     True                  6         amigo54.0.2840.191   
19999                     True                  5                firefox39.0   

                               predicted_browser_version  \
                                                           
0      [chrome55.0.2883.87, chrome56.0.2924.87, chrom...   
1      [chrome36.0.1985.125, chrome39.0.2171.71, chro...   
2      [chrome55.0.2883.87, chrome56.0.2924.87, chrom...   
3                                         [firefox7.0.1]   
4      [amigo47.5.2526.115, chrome49.0.2623.112, chro...   
...                                                  ...   
19995  [chrome55.0.2883.87, chrome56.0.2924.87, chrom...   
19996  [chrome55.0.2883.87, chrome56.0.2924.87, chrom...   
19997  [firefox12.0, firefox15.0.1, firefox38.0, fire...   
19998  [chrome55.0.2883.87, chrome56.0.2924.87, chrom...   
19999  [firefox12.0, firefox15.0.1, firefox38.0, fire...   

      browser_version_correctness browser_version_count platform  \
                                                                   
0                            True                    12  windows   
1                            True                    19  windows   
2                            True                    12  windows   
3                            True                     1  windows   
4                            True                    11  android   
...                           ...                   ...      ...   
19995                        True                    12  windows   
19996                       False                    12  windows   
19997                       False                    13  windows   
19998                       False                    12  windows   
19999                       False                    13  windows   

                                   predicted_platform platform_correctness  \
                     

In [145]:
print('Platform accuracy: {}'.format(sum(compare_frame.platform_version_correctness)/20000))
print('Conditional accuracy: {}'.format(sum(compare_frame.browser_name_correctness[0] & compare_frame.browser_version_correctness & compare_frame.platform_correctness & compare_frame.platform_version_correctness)/20000))

Platform accuracy: 0.9445
Conditional accuracy: 0.79785


### Save and load compare dataframe

In [101]:
np.save('df/compare_dummy_classification.npy', np.array(compare_frame))
print("Compare frame was saved with size(Bytes): {}".format(os.stat('df/compare_dummy_classification.npy').st_size))

Compare frame was saved with size(Bytes): 5099513


In [102]:
compare_frame2 = pd.DataFrame(np.load('df/compare_dummy_classification.npy').tolist())

In [103]:
compare_frame2.columns = [
    'browser_name', 
    'predicted_browser_name', 
    'browser_name_correctness', 
    'browser_name_count', 
    'browser_version', 
    'predicted_browser_version', 
    'browser_version_correctness', 
    'browser_version_count', 
    'platform', 
    'predicted_platform', 
    'platform_correctness', 
    'platform_count',
    'platform_version', 
    'predicted_platform_version', 
    'platform_version_correctness', 
    'platform_version_count'
]

# Conclusion (Light Threshold)

|             Layer            | DecisionTree | Thresolded Logit(0.024072) | Thresolded Logit(0.00075) | Thresolded Logit(0.01) |
|---------------------------------------------------------------------|--------------|---------|---------|---------|
| Layer 0 (browser_name)                                              |      0.79    | 0.9349  | 0.97175 | 0.96125 |
| Layer 1 (browser_name, browser_version)                             |      0.69    | 0.76575 | 0.94205 | 0.8141  |
| Layer 2 (browser_name, browser_version, platform)                   |      0.64    | 0.75885 | 0.94055 | 0.80905 |
| Layer 3 (browser_name, browser_version, platform, platform_version) |      0.65    |  0.735  |  0.9391 | 0.79785 |